<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.280">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Melanie Dube, Scientist">

<title>Intro to R for Bioinformatics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="Replication_Melanie_Dube_files/libs/clipboard/clipboard.min.js"></script>
<script src="Replication_Melanie_Dube_files/libs/quarto-html/quarto.js"></script>
<script src="Replication_Melanie_Dube_files/libs/quarto-html/popper.min.js"></script>
<script src="Replication_Melanie_Dube_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Replication_Melanie_Dube_files/libs/quarto-html/anchor.min.js"></script>
<link href="Replication_Melanie_Dube_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Replication_Melanie_Dube_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Replication_Melanie_Dube_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Replication_Melanie_Dube_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Replication_Melanie_Dube_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Intro to R for Bioinformatics</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Melanie Dube, Scientist </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="intro-to-bioinformatics" class="level1">
<h1>Intro to Bioinformatics</h1>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<section id="activity-objectives" class="level3">
<h3 class="anchored" data-anchor-id="activity-objectives">Activity Objectives</h3>
<p>Below are the listed objectives for this new notebook:</p>
<ul>
<li><p>Use the arrow (<code>&lt;-</code>) operator to store objects in variables.</p></li>
<li><p>Use R’s <code>sample()</code> function to create a random “genome” for the purpose of testing functions.</p></li>
<li><p>Build functions to process and analyze genetic data.</p></li>
<li><p>Learn about <code>for</code> loops and implement them when necessary.</p></li>
<li><p>Read in a real bacterial genome from a text file.</p></li>
<li><p>Analyse the genome using the ideas and tools developed in the first part of the notebook.</p></li>
</ul>
</section>
<section id="loading-packages" class="level3">
<h3 class="anchored" data-anchor-id="loading-packages">Loading Packages</h3>
<p>First code chunk always loads needed packages (seen below).</p>
</section>
</section>
<section id="challenge-1" class="level2">
<h2 class="anchored" data-anchor-id="challenge-1">Challenge 1</h2>
<p>The first challenge is to create a list with the four nucleotides. This is done with the <code>c()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span> )</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>nucleotides</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "C" "G" "T"</code></pre>
</div>
</div>
</section>
<section id="challenge-2" class="level2">
<h2 class="anchored" data-anchor-id="challenge-2">Challenge 2</h2>
<p>The second challenge is to adapt a code to create a random strand of 15 nucleotides. This should be called <code>randGenome</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>numNucleotides <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> numNucleotides, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "C" "T" "G" "T" "C" "T" "A" "T" "T" "C" "G" "T" "C" "G" "A"</code></pre>
</div>
</div>
<p>The block above creates the list for the random genome, and the block below will collapse it and make it more readable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTGTCTATTCGTCGA"</code></pre>
</div>
</div>
<p>Before comparing my block with a someone else, I predict that another classmate’s will be a different random order. This is because we are instructing R to generate a random code, therefore it should be different every time we run it.</p>
</section>
<section id="challenge-3" class="level2">
<h2 class="anchored" data-anchor-id="challenge-3">Challenge 3</h2>
<p>The third challenge is to generate a random genome that is 1500 nucleotides long and collapse it down to a single string.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>numNucleotidesChal3 <span class="ot">&lt;-</span> <span class="dv">1500</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>randGenomeChal3 <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> numNucleotidesChal3, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>randGenomeChal3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [1] "G" "C" "G" "A" "A" "T" "C" "T" "A" "A" "T" "A" "A" "A" "C" "A" "C" "G"
  [19] "A" "A" "C" "A" "G" "C" "T" "A" "C" "C" "G" "C" "A" "T" "C" "G" "T" "A"
  [37] "G" "A" "A" "A" "A" "A" "A" "C" "C" "G" "G" "G" "T" "A" "T" "C" "A" "A"
  [55] "T" "G" "C" "C" "T" "A" "G" "C" "G" "T" "G" "T" "T" "T" "C" "T" "A" "G"
  [73] "G" "G" "T" "T" "C" "A" "T" "A" "G" "G" "T" "T" "A" "G" "G" "A" "G" "A"
  [91] "A" "G" "A" "T" "G" "A" "G" "G" "C" "G" "G" "C" "T" "G" "T" "A" "T" "C"
 [109] "T" "T" "T" "C" "A" "T" "A" "G" "C" "G" "A" "T" "A" "G" "A" "G" "T" "G"
 [127] "G" "G" "T" "C" "A" "G" "A" "A" "T" "T" "C" "A" "A" "A" "C" "A" "T" "G"
 [145] "C" "C" "C" "C" "T" "C" "G" "A" "G" "T" "G" "G" "G" "A" "G" "C" "C" "C"
 [163] "T" "T" "C" "A" "A" "G" "T" "T" "T" "A" "A" "G" "A" "C" "C" "C" "C" "C"
 [181] "T" "A" "A" "A" "A" "C" "T" "G" "T" "C" "G" "G" "G" "C" "A" "A" "G" "A"
 [199] "T" "A" "T" "T" "T" "A" "A" "T" "G" "T" "G" "A" "C" "T" "T" "T" "A" "C"
 [217] "G" "T" "C" "T" "A" "C" "A" "T" "A" "T" "T" "A" "C" "G" "A" "A" "T" "T"
 [235] "C" "G" "A" "G" "G" "C" "A" "T" "G" "G" "A" "C" "C" "A" "C" "C" "G" "C"
 [253] "A" "G" "C" "T" "G" "A" "G" "G" "G" "T" "A" "A" "A" "A" "G" "T" "T" "G"
 [271] "G" "G" "C" "T" "A" "A" "G" "A" "G" "T" "C" "T" "A" "G" "A" "G" "G" "G"
 [289] "T" "G" "C" "A" "A" "A" "G" "T" "T" "T" "T" "T" "T" "G" "C" "T" "G" "T"
 [307] "A" "A" "C" "A" "A" "A" "G" "T" "C" "C" "T" "A" "A" "A" "C" "T" "T" "A"
 [325] "T" "T" "T" "T" "G" "C" "T" "T" "G" "G" "G" "C" "A" "C" "G" "C" "C" "A"
 [343] "T" "A" "G" "A" "C" "G" "A" "C" "T" "G" "A" "T" "T" "T" "C" "G" "A" "A"
 [361] "G" "G" "C" "C" "C" "A" "G" "G" "T" "A" "A" "C" "T" "C" "G" "A" "G" "A"
 [379] "G" "G" "C" "G" "C" "G" "A" "G" "T" "T" "A" "G" "C" "C" "T" "C" "A" "T"
 [397] "C" "A" "A" "C" "C" "G" "T" "A" "T" "C" "C" "C" "T" "T" "A" "A" "G" "T"
 [415] "T" "G" "G" "C" "A" "A" "T" "C" "T" "C" "A" "A" "G" "C" "T" "A" "T" "T"
 [433] "A" "C" "T" "A" "A" "T" "T" "T" "C" "C" "C" "C" "A" "A" "G" "C" "T" "A"
 [451] "A" "G" "G" "T" "T" "C" "G" "T" "C" "A" "C" "A" "A" "T" "T" "A" "A" "C"
 [469] "C" "T" "C" "G" "G" "C" "T" "G" "A" "T" "G" "C" "G" "T" "G" "G" "C" "G"
 [487] "T" "C" "G" "A" "G" "G" "G" "G" "G" "C" "T" "A" "A" "T" "T" "C" "T" "G"
 [505] "T" "C" "A" "G" "G" "G" "C" "C" "A" "G" "T" "T" "A" "T" "G" "T" "C" "G"
 [523] "C" "A" "G" "C" "G" "T" "A" "G" "C" "C" "G" "T" "C" "G" "C" "A" "T" "A"
 [541] "A" "A" "T" "T" "T" "A" "T" "T" "C" "T" "G" "G" "T" "C" "A" "G" "A" "C"
 [559] "C" "G" "A" "G" "T" "C" "T" "G" "T" "T" "A" "C" "C" "T" "G" "C" "G" "T"
 [577] "G" "A" "C" "G" "A" "T" "T" "C" "A" "T" "A" "C" "T" "A" "T" "C" "A" "G"
 [595] "C" "A" "C" "G" "A" "T" "A" "G" "G" "G" "G" "G" "C" "T" "C" "C" "T" "C"
 [613] "G" "T" "A" "A" "A" "C" "C" "A" "T" "A" "A" "A" "G" "T" "C" "G" "T" "C"
 [631] "G" "G" "G" "C" "C" "G" "T" "T" "G" "G" "G" "T" "A" "G" "G" "T" "G" "T"
 [649] "G" "C" "C" "T" "T" "T" "T" "A" "A" "T" "T" "A" "G" "A" "A" "C" "T" "A"
 [667] "A" "C" "C" "C" "G" "A" "A" "C" "G" "T" "T" "C" "T" "T" "A" "C" "G" "T"
 [685] "A" "G" "T" "C" "T" "G" "G" "C" "T" "A" "A" "A" "G" "G" "G" "G" "C" "G"
 [703] "A" "T" "G" "A" "T" "C" "A" "T" "A" "G" "C" "C" "A" "A" "G" "T" "G" "C"
 [721] "A" "A" "G" "G" "G" "C" "A" "G" "C" "T" "A" "G" "C" "G" "G" "T" "G" "T"
 [739] "A" "A" "T" "G" "T" "T" "C" "G" "A" "G" "C" "A" "C" "C" "T" "C" "T" "G"
 [757] "A" "A" "A" "G" "G" "T" "G" "A" "T" "T" "A" "C" "A" "A" "A" "C" "C" "T"
 [775] "G" "A" "A" "T" "T" "T" "C" "A" "G" "G" "C" "A" "T" "A" "A" "T" "G" "A"
 [793] "A" "C" "C" "C" "G" "C" "A" "T" "A" "T" "C" "C" "G" "G" "G" "C" "C" "A"
 [811] "G" "C" "A" "C" "T" "T" "G" "T" "G" "C" "G" "A" "G" "T" "T" "A" "A" "A"
 [829] "C" "C" "A" "A" "A" "T" "T" "T" "A" "T" "C" "T" "G" "G" "G" "G" "G" "A"
 [847] "C" "A" "T" "G" "C" "T" "C" "C" "A" "A" "T" "G" "C" "A" "C" "C" "A" "C"
 [865] "T" "A" "C" "C" "C" "A" "T" "C" "C" "T" "G" "T" "A" "T" "C" "C" "C" "G"
 [883] "A" "T" "T" "C" "T" "G" "G" "G" "G" "A" "A" "C" "G" "C" "A" "C" "A" "C"
 [901] "T" "C" "T" "G" "T" "A" "T" "T" "G" "T" "G" "T" "A" "A" "C" "G" "A" "A"
 [919] "G" "A" "G" "G" "T" "A" "A" "A" "C" "T" "T" "T" "G" "C" "C" "C" "A" "T"
 [937] "T" "T" "C" "T" "G" "T" "A" "G" "C" "C" "G" "G" "C" "C" "T" "G" "C" "G"
 [955] "T" "G" "T" "C" "A" "C" "A" "A" "A" "T" "T" "C" "G" "A" "T" "G" "T" "A"
 [973] "C" "A" "C" "A" "G" "C" "A" "A" "G" "A" "A" "A" "A" "C" "C" "A" "G" "G"
 [991] "A" "T" "T" "G" "G" "T" "G" "C" "A" "C" "C" "T" "G" "A" "C" "C" "T" "A"
[1009] "A" "G" "C" "G" "G" "A" "A" "C" "T" "A" "C" "C" "A" "T" "A" "A" "T" "C"
[1027] "T" "A" "G" "G" "A" "G" "T" "C" "A" "C" "A" "C" "T" "T" "A" "A" "T" "C"
[1045] "A" "G" "A" "C" "C" "G" "A" "T" "A" "C" "A" "T" "A" "C" "A" "A" "T" "C"
[1063] "C" "T" "G" "C" "G" "G" "G" "C" "G" "G" "C" "C" "C" "A" "C" "A" "G" "G"
[1081] "C" "C" "T" "T" "C" "G" "C" "T" "C" "G" "T" "A" "A" "C" "A" "T" "G" "C"
[1099] "A" "A" "C" "A" "C" "T" "T" "C" "T" "G" "G" "G" "G" "T" "A" "A" "A" "T"
[1117] "T" "T" "T" "A" "G" "G" "C" "G" "C" "C" "G" "T" "G" "T" "T" "T" "G" "A"
[1135] "A" "G" "T" "T" "C" "C" "G" "T" "A" "G" "T" "A" "C" "C" "A" "C" "G" "A"
[1153] "T" "G" "G" "T" "A" "A" "G" "T" "G" "G" "G" "G" "T" "G" "T" "T" "T" "T"
[1171] "C" "G" "T" "G" "A" "A" "C" "A" "G" "G" "C" "G" "C" "G" "T" "C" "A" "A"
[1189] "C" "T" "A" "G" "T" "G" "T" "T" "A" "A" "T" "G" "G" "C" "A" "A" "C" "G"
[1207] "G" "C" "A" "G" "C" "A" "G" "T" "G" "C" "A" "T" "T" "T" "T" "T" "C" "C"
[1225] "G" "T" "G" "A" "A" "A" "G" "A" "T" "C" "A" "A" "T" "A" "C" "A" "A" "A"
[1243] "A" "A" "T" "C" "A" "C" "C" "A" "G" "C" "T" "T" "G" "C" "A" "C" "G" "C"
[1261] "G" "G" "G" "G" "G" "C" "A" "T" "T" "C" "T" "T" "C" "T" "C" "A" "A" "C"
[1279] "T" "A" "C" "G" "T" "C" "G" "A" "C" "T" "A" "A" "T" "A" "T" "C" "C" "G"
[1297] "G" "G" "G" "A" "G" "A" "A" "C" "C" "G" "A" "T" "T" "C" "T" "G" "C" "G"
[1315] "G" "G" "C" "A" "A" "T" "G" "T" "A" "T" "A" "A" "C" "T" "C" "C" "A" "A"
[1333] "T" "G" "C" "T" "C" "A" "G" "C" "G" "A" "A" "T" "C" "T" "C" "C" "T" "A"
[1351] "C" "G" "T" "T" "T" "A" "C" "T" "G" "T" "G" "A" "T" "A" "C" "T" "A" "C"
[1369] "G" "A" "A" "T" "T" "G" "A" "A" "A" "A" "A" "C" "G" "A" "T" "C" "C" "G"
[1387] "A" "A" "G" "G" "G" "G" "G" "A" "G" "C" "T" "A" "A" "C" "T" "C" "G" "C"
[1405] "G" "A" "T" "G" "A" "C" "A" "A" "T" "C" "C" "C" "C" "T" "A" "G" "A" "A"
[1423] "C" "C" "C" "C" "C" "A" "A" "A" "A" "T" "G" "A" "A" "A" "G" "T" "C" "G"
[1441] "T" "G" "T" "C" "C" "T" "T" "A" "A" "C" "A" "A" "T" "C" "G" "C" "G" "T"
[1459] "A" "A" "A" "A" "T" "T" "G" "C" "C" "C" "A" "C" "A" "A" "C" "C" "A" "C"
[1477] "T" "C" "A" "A" "C" "C" "C" "A" "C" "C" "G" "T" "G" "C" "G" "A" "T" "G"
[1495] "A" "C" "T" "A" "A" "T"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(randGenomeChal3, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GCGAATCTAATAAACACGAACAGCTACCGCATCGTAGAAAAAACCGGGTATCAATGCCTAGCGTGTTTCTAGGGTTCATAGGTTAGGAGAAGATGAGGCGGCTGTATCTTTCATAGCGATAGAGTGGGTCAGAATTCAAACATGCCCCTCGAGTGGGAGCCCTTCAAGTTTAAGACCCCCTAAAACTGTCGGGCAAGATATTTAATGTGACTTTACGTCTACATATTACGAATTCGAGGCATGGACCACCGCAGCTGAGGGTAAAAGTTGGGCTAAGAGTCTAGAGGGTGCAAAGTTTTTTGCTGTAACAAAGTCCTAAACTTATTTTGCTTGGGCACGCCATAGACGACTGATTTCGAAGGCCCAGGTAACTCGAGAGGCGCGAGTTAGCCTCATCAACCGTATCCCTTAAGTTGGCAATCTCAAGCTATTACTAATTTCCCCAAGCTAAGGTTCGTCACAATTAACCTCGGCTGATGCGTGGCGTCGAGGGGGCTAATTCTGTCAGGGCCAGTTATGTCGCAGCGTAGCCGTCGCATAAATTTATTCTGGTCAGACCGAGTCTGTTACCTGCGTGACGATTCATACTATCAGCACGATAGGGGGCTCCTCGTAAACCATAAAGTCGTCGGGCCGTTGGGTAGGTGTGCCTTTTAATTAGAACTAACCCGAACGTTCTTACGTAGTCTGGCTAAAGGGGCGATGATCATAGCCAAGTGCAAGGGCAGCTAGCGGTGTAATGTTCGAGCACCTCTGAAAGGTGATTACAAACCTGAATTTCAGGCATAATGAACCCGCATATCCGGGCCAGCACTTGTGCGAGTTAAACCAAATTTATCTGGGGGACATGCTCCAATGCACCACTACCCATCCTGTATCCCGATTCTGGGGAACGCACACTCTGTATTGTGTAACGAAGAGGTAAACTTTGCCCATTTCTGTAGCCGGCCTGCGTGTCACAAATTCGATGTACACAGCAAGAAAACCAGGATTGGTGCACCTGACCTAAGCGGAACTACCATAATCTAGGAGTCACACTTAATCAGACCGATACATACAATCCTGCGGGCGGCCCACAGGCCTTCGCTCGTAACATGCAACACTTCTGGGGTAAATTTTAGGCGCCGTGTTTGAAGTTCCGTAGTACCACGATGGTAAGTGGGGTGTTTTCGTGAACAGGCGCGTCAACTAGTGTTAATGGCAACGGCAGCAGTGCATTTTTCCGTGAAAGATCAATACAAAAATCACCAGCTTGCACGCGGGGGCATTCTTCTCAACTACGTCGACTAATATCCGGGGAGAACCGATTCTGCGGGCAATGTATAACTCCAATGCTCAGCGAATCTCCTACGTTTACTGTGATACTACGAATTGAAAAACGATCCGAAGGGGGAGCTAACTCGCGATGACAATCCCCTAGAACCCCCAAAATGAAAGTCGTGTCCTTAACAATCGCGTAAAATTGCCCACAACCACTCAACCCACCGTGCGATGACTAAT"</code></pre>
</div>
</div>
<p>The block above successfully creates the 1500 unit random genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>genomeLength <span class="ot">&lt;-</span> <span class="dv">1500</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>randGenome4 <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> genomeLength, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>randGenome4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [1] "T" "G" "G" "A" "A" "T" "C" "T" "T" "T" "A" "A" "T" "G" "T" "T" "C" "C"
  [19] "G" "A" "C" "C" "T" "T" "T" "G" "A" "G" "C" "C" "T" "C" "A" "A" "G" "T"
  [37] "T" "G" "G" "A" "C" "T" "A" "C" "C" "G" "C" "G" "G" "C" "C" "C" "C" "C"
  [55] "G" "A" "C" "C" "G" "A" "G" "A" "G" "C" "T" "A" "A" "C" "T" "C" "T" "A"
  [73] "A" "A" "T" "G" "A" "T" "G" "C" "A" "A" "G" "C" "A" "G" "C" "G" "C" "G"
  [91] "T" "G" "C" "C" "G" "A" "G" "T" "C" "T" "C" "T" "G" "T" "G" "G" "A" "C"
 [109] "A" "G" "G" "A" "T" "C" "A" "C" "A" "T" "T" "G" "G" "A" "T" "G" "T" "G"
 [127] "A" "G" "C" "C" "A" "A" "G" "A" "T" "C" "T" "C" "G" "A" "G" "A" "T" "G"
 [145] "A" "A" "A" "A" "A" "C" "A" "A" "G" "T" "A" "G" "G" "T" "T" "C" "T" "C"
 [163] "T" "A" "T" "C" "G" "C" "T" "G" "G" "C" "T" "C" "C" "A" "A" "T" "C" "A"
 [181] "T" "T" "T" "T" "T" "A" "T" "C" "C" "T" "T" "T" "G" "C" "T" "G" "C" "A"
 [199] "T" "A" "C" "T" "T" "A" "T" "G" "G" "C" "C" "C" "T" "C" "C" "A" "A" "C"
 [217] "C" "T" "T" "G" "C" "G" "A" "A" "T" "T" "G" "C" "G" "G" "C" "A" "T" "T"
 [235] "A" "C" "G" "T" "T" "C" "T" "A" "T" "G" "C" "A" "C" "G" "A" "A" "A" "G"
 [253] "A" "G" "G" "C" "A" "C" "G" "A" "G" "A" "T" "G" "T" "A" "T" "A" "T" "C"
 [271] "A" "C" "T" "C" "T" "T" "C" "A" "G" "C" "T" "A" "T" "C" "T" "G" "A" "C"
 [289] "A" "C" "A" "G" "G" "T" "A" "G" "G" "C" "A" "T" "T" "G" "G" "A" "T" "G"
 [307] "T" "C" "A" "G" "C" "G" "T" "A" "C" "G" "A" "C" "C" "G" "G" "G" "G" "T"
 [325] "A" "G" "G" "C" "A" "A" "C" "C" "C" "C" "T" "T" "C" "T" "G" "T" "T" "G"
 [343] "C" "C" "C" "C" "G" "C" "T" "G" "G" "C" "C" "G" "G" "A" "T" "G" "C" "C"
 [361] "A" "A" "G" "G" "T" "G" "T" "T" "T" "T" "A" "C" "A" "T" "C" "G" "G" "T"
 [379] "A" "T" "T" "T" "T" "A" "A" "A" "G" "A" "T" "G" "T" "G" "A" "A" "C" "T"
 [397] "T" "A" "A" "G" "T" "A" "A" "C" "C" "T" "A" "C" "T" "T" "A" "G" "C" "T"
 [415] "G" "A" "C" "G" "G" "G" "A" "A" "G" "G" "T" "G" "C" "A" "C" "A" "T" "G"
 [433] "T" "A" "T" "G" "T" "G" "T" "G" "A" "C" "T" "C" "T" "A" "C" "A" "C" "C"
 [451] "A" "A" "G" "A" "T" "G" "C" "A" "C" "T" "T" "A" "G" "G" "C" "A" "T" "C"
 [469] "A" "A" "T" "A" "A" "A" "A" "G" "T" "T" "G" "C" "C" "G" "G" "T" "T" "T"
 [487] "G" "T" "A" "A" "T" "C" "C" "T" "T" "G" "A" "C" "A" "G" "T" "A" "A" "T"
 [505] "C" "G" "A" "G" "A" "T" "A" "A" "T" "T" "A" "C" "T" "T" "G" "C" "G" "G"
 [523] "G" "C" "A" "C" "A" "T" "A" "A" "C" "C" "T" "A" "C" "T" "C" "G" "G" "T"
 [541] "T" "C" "G" "T" "C" "C" "C" "C" "C" "T" "T" "A" "G" "C" "G" "T" "A" "C"
 [559] "G" "G" "G" "G" "G" "G" "G" "T" "G" "G" "G" "G" "A" "A" "C" "C" "A" "T"
 [577] "C" "A" "G" "C" "G" "T" "T" "C" "G" "T" "A" "T" "G" "T" "T" "A" "G" "T"
 [595] "C" "C" "T" "C" "C" "G" "G" "C" "A" "T" "T" "T" "T" "T" "G" "G" "T" "C"
 [613] "A" "C" "G" "G" "C" "C" "C" "T" "C" "C" "A" "T" "G" "A" "A" "A" "T" "A"
 [631] "C" "A" "T" "A" "C" "T" "A" "G" "C" "G" "T" "G" "C" "C" "A" "T" "T" "C"
 [649] "C" "G" "G" "T" "C" "A" "A" "A" "T" "G" "G" "C" "C" "C" "T" "G" "C" "G"
 [667] "A" "A" "G" "G" "A" "T" "G" "C" "G" "G" "T" "G" "C" "G" "G" "T" "A" "A"
 [685] "G" "C" "G" "T" "G" "T" "G" "G" "G" "C" "C" "A" "T" "A" "C" "T" "T" "G"
 [703] "G" "G" "C" "A" "G" "A" "T" "T" "G" "A" "G" "T" "T" "A" "T" "A" "A" "G"
 [721] "A" "A" "A" "C" "A" "G" "G" "C" "A" "A" "A" "T" "T" "G" "G" "G" "A" "A"
 [739] "C" "T" "A" "G" "T" "G" "C" "G" "G" "C" "A" "C" "A" "T" "G" "T" "A" "C"
 [757] "A" "G" "C" "T" "C" "G" "C" "A" "C" "T" "T" "C" "C" "T" "G" "G" "G" "G"
 [775] "G" "G" "C" "G" "A" "C" "T" "A" "C" "C" "G" "A" "A" "T" "T" "C" "C" "T"
 [793] "A" "G" "T" "A" "T" "G" "T" "C" "T" "G" "A" "G" "T" "A" "G" "C" "G" "T"
 [811] "A" "C" "C" "G" "G" "G" "C" "C" "A" "G" "C" "T" "T" "T" "G" "G" "G" "T"
 [829] "C" "C" "T" "C" "T" "T" "A" "C" "C" "T" "T" "G" "T" "A" "A" "T" "G" "G"
 [847] "G" "A" "T" "C" "G" "G" "C" "C" "T" "T" "A" "T" "G" "G" "G" "A" "T" "G"
 [865] "C" "C" "A" "T" "G" "A" "C" "G" "A" "C" "A" "C" "T" "G" "C" "G" "G" "G"
 [883] "A" "G" "G" "T" "T" "A" "A" "T" "A" "A" "G" "T" "G" "T" "C" "C" "G" "C"
 [901] "G" "A" "T" "A" "A" "A" "G" "C" "C" "T" "C" "G" "T" "G" "G" "T" "C" "G"
 [919] "A" "A" "A" "T" "C" "T" "G" "T" "T" "A" "C" "G" "T" "G" "C" "T" "C" "T"
 [937] "T" "C" "T" "G" "T" "C" "A" "T" "T" "T" "G" "C" "C" "A" "T" "A" "T" "T"
 [955] "G" "G" "T" "A" "A" "A" "G" "T" "T" "C" "A" "A" "G" "A" "G" "C" "C" "A"
 [973] "C" "C" "C" "T" "A" "T" "C" "T" "A" "T" "T" "G" "A" "C" "T" "C" "C" "T"
 [991] "A" "G" "A" "T" "C" "T" "C" "C" "C" "A" "A" "A" "T" "G" "T" "A" "T" "C"
[1009] "C" "T" "C" "C" "A" "A" "C" "A" "G" "T" "A" "C" "T" "T" "C" "A" "C" "C"
[1027] "T" "G" "G" "A" "T" "G" "G" "T" "A" "A" "C" "C" "C" "C" "A" "C" "C" "T"
[1045] "C" "A" "A" "T" "C" "A" "C" "C" "A" "C" "T" "T" "A" "G" "G" "T" "C" "T"
[1063] "C" "C" "A" "G" "T" "A" "C" "T" "T" "T" "T" "G" "A" "G" "C" "C" "G" "C"
[1081] "C" "T" "A" "C" "G" "C" "G" "C" "T" "C" "T" "C" "G" "A" "A" "A" "T" "A"
[1099] "A" "G" "C" "T" "A" "A" "G" "C" "T" "A" "G" "C" "T" "A" "C" "G" "G" "T"
[1117] "C" "C" "G" "T" "G" "T" "T" "A" "G" "G" "C" "G" "A" "A" "A" "A" "A" "T"
[1135] "G" "A" "A" "T" "C" "C" "C" "T" "T" "T" "C" "C" "A" "G" "C" "G" "C" "T"
[1153] "A" "C" "C" "A" "A" "G" "T" "G" "T" "A" "G" "T" "C" "G" "G" "A" "T" "G"
[1171] "C" "G" "G" "T" "T" "G" "A" "C" "T" "T" "A" "T" "G" "C" "C" "T" "G" "C"
[1189] "C" "C" "G" "G" "T" "G" "A" "G" "C" "A" "G" "A" "A" "T" "T" "T" "A" "A"
[1207] "A" "T" "C" "G" "C" "T" "G" "T" "A" "C" "C" "A" "A" "T" "C" "A" "G" "C"
[1225] "A" "C" "T" "A" "C" "C" "A" "C" "A" "A" "G" "G" "C" "G" "A" "A" "C" "G"
[1243] "G" "G" "G" "C" "A" "G" "C" "C" "A" "C" "G" "A" "C" "T" "G" "G" "C" "G"
[1261] "C" "C" "A" "C" "A" "A" "G" "C" "A" "T" "G" "G" "A" "A" "G" "G" "A" "G"
[1279] "T" "G" "T" "C" "G" "G" "G" "C" "G" "A" "A" "A" "T" "C" "A" "G" "G" "G"
[1297] "C" "C" "G" "C" "A" "T" "T" "A" "A" "G" "C" "C" "G" "T" "T" "T" "T" "A"
[1315] "T" "C" "A" "C" "G" "G" "G" "T" "A" "A" "A" "G" "A" "G" "G" "T" "C" "A"
[1333] "C" "A" "T" "G" "T" "G" "C" "A" "C" "A" "G" "A" "G" "C" "C" "A" "G" "A"
[1351] "T" "A" "G" "T" "C" "G" "G" "C" "A" "G" "T" "G" "A" "A" "G" "G" "G" "C"
[1369] "A" "C" "A" "A" "C" "G" "C" "T" "T" "T" "T" "T" "C" "G" "G" "C" "A" "G"
[1387] "G" "A" "A" "G" "A" "C" "C" "C" "T" "T" "G" "C" "A" "C" "C" "T" "T" "T"
[1405] "A" "T" "C" "A" "C" "A" "C" "C" "G" "C" "G" "G" "G" "T" "C" "A" "T" "C"
[1423] "A" "C" "C" "G" "G" "C" "C" "A" "A" "A" "G" "C" "A" "G" "T" "C" "G" "G"
[1441] "C" "G" "G" "T" "C" "A" "A" "C" "G" "T" "T" "A" "A" "A" "A" "T" "T" "A"
[1459] "T" "T" "G" "C" "A" "T" "G" "T" "C" "T" "A" "G" "G" "T" "A" "A" "A" "G"
[1477] "C" "A" "G" "G" "A" "C" "C" "A" "G" "A" "G" "A" "G" "A" "C" "G" "T" "C"
[1495] "G" "A" "A" "A" "T" "C"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>randGenome4 <span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Will need to compare this to others when able to do so.</p>
</section>
<section id="challenge-4" class="level2">
<h2 class="anchored" data-anchor-id="challenge-4">Challenge 4</h2>
<p>The fourth challenge is to use the <code>set.seed(215)</code> command to create a random genome of 100 nucleotides in a single string.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>genomeLength <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> genomeLength, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "T" "G" "G" "A" "A" "T" "C" "T" "T" "T" "A" "A" "T" "G" "T" "T" "C" "C"
 [19] "G" "A" "C" "C" "T" "T" "T" "G" "A" "G" "C" "C" "T" "C" "A" "A" "G" "T"
 [37] "T" "G" "G" "A" "C" "T" "A" "C" "C" "G" "C" "G" "G" "C" "C" "C" "C" "C"
 [55] "G" "A" "C" "C" "G" "A" "G" "A" "G" "C" "T" "A" "A" "C" "T" "C" "T" "A"
 [73] "A" "A" "T" "G" "A" "T" "G" "C" "A" "A" "G" "C" "A" "G" "C" "G" "C" "G"
 [91] "T" "G" "C" "C" "G" "A" "G" "T" "C" "T"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TGGAATCTTTAATGTTCCGACCTTTGAGCCTCAAGTTGGACTACCGCGGCCCCCGACCGAGAGCTAACTCTAAATGATGCAAGCAGCGCGTGCCGAGTCT"</code></pre>
</div>
</div>
<p>The number of Adenine (A) in this sequence: <strong>23</strong> (need to verify with others)</p>
<section id="loops" class="level3">
<h3 class="anchored" data-anchor-id="loops">Loops</h3>
<p>Basic idea of using a <code>for</code> loop function:</p>
<ol type="1">
<li><p>Set an iterator (this may just be a counter or it may be floating over elements in an existing list).</p></li>
<li><p>Provide the instructions to be run each time through the loop.</p></li>
</ol>
<p>Practicing loops:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>mySum <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>){</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  mySum <span class="ot">&lt;-</span> mySum <span class="sc">+</span> i</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(mySum)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1
[1] 3
[1] 6
[1] 10
[1] 15
[1] 21
[1] 28
[1] 36
[1] 45
[1] 55</code></pre>
</div>
</div>
</section>
</section>
<section id="challenge-5" class="level2">
<h2 class="anchored" data-anchor-id="challenge-5">Challenge 5</h2>
<p>The fifth challenge is to initialize a variable called <code>myProduct</code> (with the value 1) and write a loop using the iterator <code>j</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>myProduct <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>){</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  myProduct <span class="ot">&lt;-</span> myProduct <span class="sc">+</span> j</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(myProduct)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2
[1] 4
[1] 7
[1] 11
[1] 16
[1] 22
[1] 29
[1] 37
[1] 46
[1] 56
[1] 67
[1] 79
[1] 92
[1] 106
[1] 121</code></pre>
</div>
</div>
<p>This code block runs the <code>for</code> loop with the created variable. Need to verify this with others when get the chance.</p>
</section>
<section id="challenge-6" class="level2">
<h2 class="anchored" data-anchor-id="challenge-6">Challenge 6</h2>
<p>The sixth challenge is to create a random genome sub string of 10 nucleotides and use the <code>paste</code> function to collapse to a single string. Then, we need to use a <code>for</code> loop to print out each nucleotide.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>numNucleotides <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> numNucleotides, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "C" "T" "G" "T" "G" "G" "A" "C" "A" "G"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTGTGGACAG"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>){</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i))</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C"
[1] "T"
[1] "G"
[1] "T"
[1] "G"
[1] "G"
[1] "A"
[1] "C"
[1] "A"
[1] "G"</code></pre>
</div>
</div>
<p>The above code block combines multiple functions to get a simple result. First, the <code>randGenome</code> function is created for 10 random nucleotides. Next, the <code>for</code> loop function creates the nucleotides into a single string, utilizing the <code>str_sub</code> command. This creates the whole strand into a single character.</p>
</section>
<section id="challenge-7" class="level2">
<h2 class="anchored" data-anchor-id="challenge-7">Challenge 7</h2>
<p>This challenge asks to count the number of occurrences of Adenine in the randGenome sequence. This is done using an adaptation of the loop function. By adding <code>Adenine_count + 1</code>, the code knows to update the adenine count every time an <code>A</code> is encountered in the <code>randGenome</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTGTGGACAG"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>randGenome<span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>Adenine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    Adenine_count <span class="ot">&lt;-</span> Adenine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Adenine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
</section>
<section id="challenge-8" class="level2">
<h2 class="anchored" data-anchor-id="challenge-8">Challenge 8</h2>
<p>In this challenge we must count the frequencies for each of the four nucleotides.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTGTGGACAG"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>randGenome<span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>Guanine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    Guanine_count <span class="ot">&lt;-</span> Guanine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Guanine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>This counts the frequency for <code>G</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTGTGGACAG"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>randGenome<span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>Cytosine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    Cytosine_count <span class="ot">&lt;-</span> Cytosine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cytosine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>This counts the frequency for <code>C</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTGTGGACAG"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>randGenome<span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>Thymine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    Thymine_count <span class="ot">&lt;-</span> Thymine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Thymine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>This counts the frequency for <code>T</code>.</p>
</section>
<section id="challenge-9" class="level2">
<h2 class="anchored" data-anchor-id="challenge-9">Challenge 9</h2>
<p>Now we need to use the same concepts to count the frequencies on a real genome with over a million nucleotides.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>vib_c <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"~/Desktop/Bioinformatics/VibrioCholerae.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="co">#vib_c</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Above is the entire genome for Vibrio Cholerae. Below is the code to count each of the nucleotides individually.</p>
<p>Counts for:</p>
<ul>
<li><p>A = <strong>293,942</strong></p></li>
<li><p>G = <strong>256,024</strong></p></li>
<li><p>C = <strong>263,573</strong></p></li>
<li><p>T = <strong>294,711</strong></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>vib_c<span class="ot">&lt;-</span> <span class="fu">paste</span>(vib_c, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>Cytosine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(vib_c)){</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    Cytosine_count <span class="ot">&lt;-</span> Cytosine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cytosine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 263573</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>vib_c<span class="ot">&lt;-</span> <span class="fu">paste</span>(vib_c, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>Adenine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(vib_c)){</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>    Adenine_count <span class="ot">&lt;-</span> Adenine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Adenine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 293942</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>vib_c<span class="ot">&lt;-</span> <span class="fu">paste</span>(vib_c, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>Guanine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(vib_c)){</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    Guanine_count <span class="ot">&lt;-</span> Guanine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Guanine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 256024</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>vib_c<span class="ot">&lt;-</span> <span class="fu">paste</span>(vib_c, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>Thymine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(vib_c)){</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    Thymine_count <span class="ot">&lt;-</span> Thymine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Thymine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 294711</code></pre>
</div>
</div>
<p>This could also all be combined into one <code>for</code> loop, to avoid running the genome so many times. Each of the components would each be repeated three times (count, if statements, and print) for each of the individual nucleotides.</p>
</section>
<section id="challenge-10" class="level2">
<h2 class="anchored" data-anchor-id="challenge-10">Challenge 10</h2>
<p>The tenth challenge is to navigate and verify the Rosalind challenge. The data is uploaded below and the number of individual nucleotides is calculated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>rosalind_dna <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"~/Desktop/Bioinformatics/rosalind_dna.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>rosalind_dna</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ATGGGGGACAGACTTAATGCTAGGTAAGTAGGCGTTAGCGCCACGCACTAGGCGTGATTGTTAGACCAAAGCCCCTTCAGGCGGCCGATCGACACTAGAATCGCCAACTCCGCGGCCGTTTTGTTATGGGGGCTGTTACGCAGTGTCCCGCTAACGGTAGGTTGAACAGATCCACAAAACAGCTCAAAATTCTCCTGTCGATTTGGCCTGAATCCATTAATGTTGTGGAAAGTACGTTCATACCGATCACGTCATTCTTAGGCAGGAAGGGTCCAAAGTCAGGGAATTATTGAGTAAGTCACGTGGGCTGCGGAACATCAATTGGTGCCCCTAGGGCTTTGTTATTATTGGGAGTGTTCGAGTTCCAGATCGGAAATATCTAACCAAATAGAAGGGGTACTCAGAAGAATTGTTTCTGTACTTTAGTGAACAGTGTTTTGTTCCAGCACTCTCTTCAGGGACTCCGTAGCAAAAGCATGGTACAGTGGCTAGGCAGTATAGTACTATTTGCGTTGAAGCTTTACGGGGTTGTTTCTGCTATCGCATCCACTGAGCACAGCCCGTTTTCTAGGGATGGGGCGCTTAATACGACAATAGTCGGGAGATCCGAGAGCCCACGGGTTCGTTGACCTAAACGTAAGCAACGTGTTAGCACATCCCTTCATGAGGGTGTTCAGGTACGAGTATCACATACGAAACAAAGGGGCAATACTGCGATTGCGGATCCTGTCCCCGACACTGTTTTGACATACCTAGACACTGTGCCTCATATTCCGAAGGGCAATCATCCAGCAGTACTCCTACGACTTTAGTTCGCTAGTGTTCGCGCCACCACTCTAGAGAACCATTATGTTCGACGCGACAAGCGGTTCAGTCCCCTTGCCGTTCGCGGACTTTTCATAGCTAATCGGCCAGCGGGTTGGTCTATCTTTATGA"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>rosalind_dna<span class="ot">&lt;-</span> <span class="fu">paste</span>(rosalind_dna, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>Thymine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>Cytosine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>Guanine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>Adenine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rosalind_dna)){</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>    Thymine_count <span class="ot">&lt;-</span> Thymine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>    Cytosine_count <span class="ot">&lt;-</span> Cytosine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a>    Guanine_count <span class="ot">&lt;-</span> Guanine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a>    Adenine_count <span class="ot">&lt;-</span> Adenine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Thymine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 247</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cytosine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 219</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Guanine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 240</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Adenine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 230</code></pre>
</div>
</div>
<p>Final Counts:</p>
<ul>
<li><p>T: 247</p></li>
<li><p>C: 219</p></li>
<li><p>G: 240</p></li>
<li><p>A: 230</p></li>
</ul>
<p>The challenge timed out before I was able to submit and verify my answers for the first data set, a second attempt is shown below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>rosalind_dna1 <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"~/Desktop/Bioinformatics/rosalind_dna1.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>rosalind_dna1<span class="ot">&lt;-</span> <span class="fu">paste</span>(rosalind_dna1, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>Thymine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>Cytosine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>Guanine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>Adenine_count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rosalind_dna1)){</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna1, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>    Thymine_count <span class="ot">&lt;-</span> Thymine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna1, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>    Cytosine_count <span class="ot">&lt;-</span> Cytosine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna1, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a>    Guanine_count <span class="ot">&lt;-</span> Guanine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_dna1, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a>    Adenine_count <span class="ot">&lt;-</span> Adenine_count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Thymine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 211</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cytosine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 214</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Guanine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 173</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Adenine_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 215</code></pre>
</div>
</div>
<p>After submitting the answer to Rosalind, I was able to verify that these were the correct totals for each nucleotide for this data set.</p>
</section>
</section>
<section id="replication-origin-part-i" class="level1">
<h1>Replication Origin: Part I</h1>
<section id="intro-functions" class="level2">
<h2 class="anchored" data-anchor-id="intro-functions">Intro: Functions</h2>
<p>This is a function which takes in two arguments, a string (<code>genomeString</code>) and a <code>nucleotide</code>. The result of the function is the frequency of the <code>nucleotide</code> within the <code>genomeString</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>nucleotide_frequency <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, <span class="at">nucleotide =</span> <span class="st">"A"</span>){</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genomeString)){</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> nucleotide){</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>      count <span class="ot">&lt;-</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(count)</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a><span class="fu">nucleotide_frequency</span>(<span class="st">"ACTTGCGGGTATCGAG"</span>, <span class="st">"G"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>This shows that there are 6 <code>guanine</code> in this string of nucleotides.</p>
</section>
<section id="challenge-1-1" class="level2">
<h2 class="anchored" data-anchor-id="challenge-1-1">Challenge 1</h2>
<p>The first challenge for this section is to use the <code>sample()</code> function to create a random genome at least 2000 in length. Then, we must count the frequency of Cytosine.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>nt_sample <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> <span class="dv">2000</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>nt_sample <span class="ot">&lt;-</span> <span class="fu">paste</span>(nt_sample, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nucleotide_frequency</span>(nt_sample, <span class="st">"C"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 487</code></pre>
</div>
</div>
</section>
<section id="challenge-2-1" class="level2">
<h2 class="anchored" data-anchor-id="challenge-2-1">Challenge 2</h2>
<p>In this challenge we must build a new function, <code>rand_genome()</code>, which uses the parameter <code>k</code> to determine how many nucleotides we want to generate in the genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>rand_genome <span class="ot">&lt;-</span> <span class="cf">function</span>(k){</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>  randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size =</span> k, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>  randGenome <span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(randGenome)</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rand_genome</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AACAGTAAAT"</code></pre>
</div>
</div>
</section>
<section id="challenge-3-1" class="level2">
<h2 class="anchored" data-anchor-id="challenge-3-1">Challenge 3</h2>
<p>In this challenge we a building a new function called <code>generate_3_mers()</code> to generate all the substrings of 3 nucleotides. Next, we want to use the <code>rand-genome()</code> function to construct a random genome of 2000 nucleotides in length, then use the <code>generate_3_mers()</code> to compute all the 3-mers in the random genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>myString <span class="ot">&lt;-</span> <span class="fu">rand_genome</span>(<span class="dv">10</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>generate_3_mers <span class="ot">&lt;-</span> <span class="cf">function</span>(myString) {</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>  list_3_mers <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nchar</span>(myString) <span class="sc">-</span> <span class="dv">2</span>)){</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>  list_3_mers <span class="ot">&lt;-</span> list_3_mers <span class="sc">%&gt;%</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(myString, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> <span class="dv">2</span>))</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_3_mers)</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>myString</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GACATTAGAA"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_3_mers</span>(myString)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GAC" "ACA" "CAT" "ATT" "TTA" "TAG" "AGA" "GAA"</code></pre>
</div>
</div>
</section>
<section id="challenge-4-1" class="level2">
<h2 class="anchored" data-anchor-id="challenge-4-1">Challenge 4</h2>
<p>For this challenge, we will create a new, updated version of the <code>generate_3_mers()</code> function so that it can generate k-mers of any length (must be less than the size of the original <code>genomeString</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>generate_k_mers <span class="ot">&lt;-</span> <span class="cf">function</span>(string, <span class="at">k =</span> <span class="dv">3</span>) {</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  list_codon <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nchar</span>(string) <span class="sc">-</span> k <span class="sc">+</span> <span class="dv">1</span>, <span class="at">by =</span> k)) {</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>    list_codon <span class="ot">&lt;-</span> list_codon <span class="sc">%&gt;%</span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">append</span>(<span class="fu">str_sub</span>(string, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_codon)</span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_k_mers</span>(<span class="fu">rand_genome</span>(<span class="dv">9</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GAC" "AAG" "ATG"</code></pre>
</div>
</div>
</section>
<section id="challenge-5-1" class="level2">
<h2 class="anchored" data-anchor-id="challenge-5-1">Challenge 5</h2>
<p>The last challenge is to create a new function which will count the number of occurrences of a particular pattern within the larger genome string.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>count_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(string, pattern){</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>  nt_matches <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(string)){</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_sub</span>(string, i, i <span class="sc">+</span> <span class="fu">str_length</span>(pattern)<span class="sc">-</span><span class="dv">1</span>) <span class="sc">==</span> pattern){</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>      nt_matches <span class="ot">=</span> nt_matches <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(nt_matches)</span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a><span class="fu">count_pattern</span>(<span class="fu">rand_genome</span>(<span class="dv">2000</span>), <span class="st">"CTG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 29</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">count_pattern</span>(<span class="st">"CGTAACTCGT"</span>, <span class="st">"CGT"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>This code successfully creates a random genome, and then counts the number of specified kmers which are identified in quotes.</p>
<section id="part-2" class="level3">
<h3 class="anchored" data-anchor-id="part-2">Part 2</h3>
<p>This part of the challenge uses data from rosalind to count the number of identified kmers. The rosalind data is loaded below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>rosalind_string <span class="ot">&lt;-</span> <span class="st">"TTAGTCCCCAGTCCCCAGTCCCCTCCAGTCCCCGCAGTCCCCCAGTCCCCAGTCCCCCAGTCCCCAGTCCCCAGTCCCCAGCCAGTCCCCACCGGTGTGGTAGTCCCCCAAGTCCCCAAGTCCCCAGTGATAACAGTCCCCTTCTCTAAGTCCCCAGTCCCCGAGTCCCCAGTTGAGTCCCCCTAGTCCCCGCCTATAGTCCCCCCACGAGTCCCCTGAAGTCCCCTGAAAGTCCCCCTGACGCAAGTCCCCTAGTCCCCCAGTCCCCAGAAGTCCCCCAGTCCCCTAAGTCCCCTAGAGTCCCCAGTCCCCGAGAGTCCCCTGTAAGTCCCCCTCAGTCCCCGGCTCGAGTCCCCGATGAGTCCCCGAGTCCCCCCGAGTCCCCGGTTAGTCCCCAAGTCCCCGAGTCCCCGAGTCCCCTGAAGTCCCCGAGTCCCCTCGAGTCCCCAGTCCCCGCTAGTCCCCCTTAGTCCCCAGTCCCCGAGTCCCCAGTCCCCAGTCCCCAAGTCCCCCGTGGAGAAGTCCCCGCAGTCCCCAGTCCCCTCGATTAGTCCCCATGCGATAGTCCCCCAGTCCCCTGAGTCCCCAGTCCCCAAGTCCCCGTTAGTCCCCGAGTCCCCAAAATTAGTCCCCGAAGTCCCCCCGTAGTCCCCTGTGAGTCCCCGAGTCCCCAGTCCCCTTACGAGTCCCCGTCCAGTCCCCTGATTATATGAGAGTCCCCTTGGAGTCCCCTAGTCCCCTAAGTCCCCAGAGTGATTCTTAGTCCCCAAGTCCCCAGTCCCCGCTAGTCCCCATAGTCCCCAGTCCCCCGCAGTCCCCCTACCTCAGTCCCCAAGTCCCCCAGTCCCCCAGTCCCCGGTATTAGTCCCCGAAGTCCCCGAGTCCCCATACTCAAGTCCCCCAGTCCCCGGATGGTAGAGTCCCCAGTCCCCTAGTCCCCCCGAGTCCCCAGTCCCCCAGTCCCCACGGCGGCTAAAGTCCCCTAGTCCCCAGTCCCCGATGCAGTCCCCAAGTCCCCAGTCCCC"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This section uses the command created above to count the number of <code>AGTCCCCAG</code> in the code, which is appears to be 25.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">count_pattern</span>(rosalind_string, <span class="st">"AGTCCCCAG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="replication-origin-part-2" class="level1">
<h1>Replication Origin: Part 2</h1>
<section id="challenge-1-2" class="level2">
<h2 class="anchored" data-anchor-id="challenge-1-2">Challenge 1</h2>
<p>The first challenge is to count the number of frequent words in a string.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>find_frequent_kmers <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, k){</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  kmers <span class="ot">&lt;-</span> <span class="fu">generate_k_mers</span>(genome, k)</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  kmers <span class="ot">&lt;-</span> <span class="fu">unique</span>(kmers)</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>  kmer_counts <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">length</span>(kmers))</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span> <span class="fu">length</span>(kmers)){</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>    kmer_counts[i]<span class="ot">&lt;-</span> <span class="fu">count_pattern</span>(genome, kmers[i])</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>max_freq <span class="ot">&lt;-</span> <span class="fu">max</span>(kmer_counts)</span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(max_freq)</span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>freq_kmers <span class="ot">&lt;-</span> kmers[kmer_counts<span class="sc">==</span>max_freq]</span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a><span class="fu">return</span>(freq_kmers)</span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>myGenome <span class="ot">&lt;-</span> <span class="fu">rand_genome</span>(<span class="dv">10</span>)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">find_frequent_kmers</span>(myGenome, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CA" "CG" "CT" "GT" "AT"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">find_frequent_kmers</span>(<span class="st">"AAATAAAG"</span>, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AAA"</code></pre>
</div>
</div>
<p>Next, we want to try to run the function we created with the data from Rosalind.</p>
<p>This is testing the function with the sample data from Rosalind.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">find_frequent_kmers</span>(<span class="st">"ACGTTGCATGTCGCATGATGCATGAGAGCT"</span>, <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GCAT" "CATG"</code></pre>
</div>
</div>
<p>This is the</p>
</section>
<section id="challenge-2-2" class="level2">
<h2 class="anchored" data-anchor-id="challenge-2-2">Challenge 2</h2>
<p>This is where we begin to think about generating the reverse complement of each of the nucleotides. Below we are building the function for the <code>reverse_complement</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>reverse_complement <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeSubString){</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  complement <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'A'</span> <span class="ot">=</span> <span class="st">'T'</span>, <span class="st">'C'</span> <span class="ot">=</span> <span class="st">'G'</span>, <span class="st">'T'</span> <span class="ot">=</span> <span class="st">'A'</span>, <span class="st">'G'</span> <span class="ot">=</span> <span class="st">'C'</span>)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  reverse_complement <span class="ot">&lt;-</span> <span class="st">''</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (nucleotide <span class="cf">in</span> <span class="fu">rev</span>(<span class="fu">strsplit</span>(genomeSubString, <span class="st">""</span>) [[<span class="dv">1</span>]])){</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>    reverse_complement <span class="ot">&lt;-</span> <span class="fu">paste</span>(reverse_complement, complement[nucleotide], <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(reverse_complement)</span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a><span class="fu">reverse_complement</span>(<span class="st">"ATTG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CAAT"</code></pre>
</div>
</div>
<p>The key to the reverse complement function is defining what we want to see when a certain nucleotide is presented (ex: A = T). This function defines that and successfully reports the reverse complement for each given nucleotide.</p>
</section>
</section>
<section id="replication-origin-part-3" class="level1">
<h1>Replication Origin: Part 3</h1>
<section id="challenge-1-3" class="level2">
<h2 class="anchored" data-anchor-id="challenge-1-3">Challenge 1</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>initialize_k_mer_dict <span class="ot">&lt;-</span> <span class="cf">function</span>(k){</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>  k_mers_dict <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">rep</span>(<span class="fu">list</span>(nucleotides), k)) <span class="sc">%&gt;%</span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unite</span>(<span class="st">"k_mers"</span>, <span class="fu">everything</span>(), <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">count =</span> <span class="dv">0</span>)</span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(k_mers_dict)</span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a>k_mers_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(<span class="dv">9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we will create the function <code>clump_finding</code> to help us determine the most frequent kmers and eventually find the origin of replication.</p>
<p>Notes: L = window size</p>
<p>k = kmer length</p>
<p>t = frequency threshold</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>initialize_k_mer_dict <span class="ot">&lt;-</span> <span class="cf">function</span>(k){</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>  k_mers_dict <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">rep</span>(<span class="fu">list</span>(nucleotides), k)) <span class="sc">%&gt;%</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unite</span>(<span class="st">"k_mers"</span>, <span class="fu">everything</span>(), <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">count =</span> <span class="dv">0</span>)</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(k_mers_dict)</span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>k_mers_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(<span class="dv">9</span>)</span>
<span id="cb103-13"><a href="#cb103-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-14"><a href="#cb103-14" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, L, k, t){</span>
<span id="cb103-15"><a href="#cb103-15" aria-hidden="true" tabindex="-1"></a>  initial_window <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">end =</span> L)</span>
<span id="cb103-16"><a href="#cb103-16" aria-hidden="true" tabindex="-1"></a>  my_kmer_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(k)</span>
<span id="cb103-17"><a href="#cb103-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span> L <span class="sc">-</span> k <span class="sc">+</span> <span class="dv">1</span>){</span>
<span id="cb103-18"><a href="#cb103-18" aria-hidden="true" tabindex="-1"></a>    curr_kmer <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(initial_window, i, i <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb103-19"><a href="#cb103-19" aria-hidden="true" tabindex="-1"></a>    curr_count<span class="ot">&lt;-</span> <span class="fu">count_pattern</span>(initial_window, curr_kmer)</span>
<span id="cb103-20"><a href="#cb103-20" aria-hidden="true" tabindex="-1"></a>    curr_row <span class="ot">&lt;-</span> <span class="fu">which</span>(my_kmer_dict<span class="sc">$</span>kmer <span class="sc">==</span> curr_kmer)</span>
<span id="cb103-21"><a href="#cb103-21" aria-hidden="true" tabindex="-1"></a>    my_kmer_dict<span class="sc">$</span>count[curr_row] <span class="ot">&lt;-</span> curr_count</span>
<span id="cb103-22"><a href="#cb103-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb103-23"><a href="#cb103-23" aria-hidden="true" tabindex="-1"></a>  candidates<span class="ot">&lt;-</span> my_kmer_dict<span class="sc">%&gt;%</span></span>
<span id="cb103-24"><a href="#cb103-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filer</span>(count<span class="sc">&gt;=</span>t)<span class="sc">%&gt;%</span></span>
<span id="cb103-25"><a href="#cb103-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(k_mer)</span>
<span id="cb103-26"><a href="#cb103-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-27"><a href="#cb103-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>(ng <span class="sc">-</span> (L <span class="sc">-</span> <span class="dv">2</span>))){</span>
<span id="cb103-28"><a href="#cb103-28" aria-hidden="true" tabindex="-1"></a>    rem_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">-</span> <span class="dv">1</span>, j <span class="sc">-</span> <span class="dv">1</span> <span class="sc">+</span> (k <span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb103-29"><a href="#cb103-29" aria-hidden="true" tabindex="-1"></a>    add_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">+</span> (L <span class="sc">-</span> <span class="dv">2</span>) <span class="sc">-</span> (k <span class="sc">-</span> <span class="dv">1</span>), j <span class="sc">+</span> (L <span class="sc">-</span> <span class="dv">2</span>))</span>
<span id="cb103-30"><a href="#cb103-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb103-31"><a href="#cb103-31" aria-hidden="true" tabindex="-1"></a>    rem_row <span class="ot">&lt;-</span> <span class="fu">which</span>(k_mers_dict<span class="sc">$</span>k_mers <span class="sc">==</span> rem_pattern)</span>
<span id="cb103-32"><a href="#cb103-32" aria-hidden="true" tabindex="-1"></a>    add_row <span class="ot">&lt;-</span> <span class="fu">which</span>(k_mers_dict<span class="sc">$</span>k_mers <span class="sc">==</span> add_pattern)</span>
<span id="cb103-33"><a href="#cb103-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-34"><a href="#cb103-34" aria-hidden="true" tabindex="-1"></a>    k_mers_dict<span class="sc">$</span>count[rem_row] <span class="ot">&lt;-</span> k_mers_dict<span class="sc">$</span>count[rem_row] <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb103-35"><a href="#cb103-35" aria-hidden="true" tabindex="-1"></a>    k_mers_dict<span class="sc">$</span>count[add_row] <span class="ot">&lt;-</span> k_mers_dict<span class="sc">$</span>count[add_row] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb103-36"><a href="#cb103-36" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-37"><a href="#cb103-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(k_mers_dict<span class="sc">$</span>count[add_row] <span class="sc">&gt;=</span> t){</span>
<span id="cb103-38"><a href="#cb103-38" aria-hidden="true" tabindex="-1"></a>      candidates <span class="ot">&lt;-</span> <span class="fu">append</span>(candidates, add_pattern)</span>
<span id="cb103-39"><a href="#cb103-39" aria-hidden="true" tabindex="-1"></a>      candidates <span class="ot">&lt;-</span> <span class="fu">unique</span>(candidates)</span>
<span id="cb103-40"><a href="#cb103-40" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb103-41"><a href="#cb103-41" aria-hidden="true" tabindex="-1"></a>  } </span>
<span id="cb103-42"><a href="#cb103-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(candidates)</span>
<span id="cb103-43"><a href="#cb103-43" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="challenge-2-3" class="level2">
<h2 class="anchored" data-anchor-id="challenge-2-3">Challenge 2</h2>
<p>Here we must implement the code that we built in challenge 1 in order to apply it to an actual genome. Here, we will find all the 9-mers corresponding to (500, 3)-clumps within the genome for E.coli. While we have built a functioning function, it is not optimized to run on a genome as large as the E. coli one when looking for 9-mers.</p>
<div class="cell">

</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>